plugins/outputs/otlp/client.go

```diff
@@ -0,0 +1,314 @@
+		return nil, err
+	}
+
+	return conn, err
+}
+
+// Selftest sends an empty request the endpoint.
+func (c *Client) Selftest(ctx context.Context) error {
+	// Loop until the context is canceled, allowing for retryable failures.
+	for {
+		conn, err := c.getConnection(ctx)
+
+		if err == nil {
+			service := metricsService.NewMetricsServiceClient(conn)
+			empty := &metricsService.ExportMetricsServiceRequest{}
+
+			_, err = service.Export(c.grpcMetadata(ctx), empty)
+			if err == nil {
+				return nil
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			if isRecoverable(err) {
+				_ = level.Info(c.logger).Log("msg", "selftest recoverable error, still trying", "err", err)
+				continue
+			}
+		}
+		return fmt.Errorf(
+			"non-recoverable failure in selftest: %s",
+			truncateErrorString(err),
+		)
+	}
+}
+
+// Store sends a batch of samples to the endpoint.
+func (c *Client) Store(req *metricsService.ExportMetricsServiceRequest) error {
+	tss := req.ResourceMetrics
+	if len(tss) == 0 {
+		// Nothing to do, return silently.
+		return nil
+	}
+
+	// Note the call to getConnection() applies its own timeout for Dial().
+	ctx := context.Background()
+	conn, err := c.getConnection(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, c.timeout)
+	defer cancel()
+
+	service := metricsService.NewMetricsServiceClient(conn)
+
+	errs := make(chan error, len(tss)/maxTimeseriesPerRequest+1)
+	var wg sync.WaitGroup
+	for i := 0; i < len(tss); i += maxTimeseriesPerRequest {
+		end := i + maxTimeseriesPerRequest
+		if end > len(tss) {
+			end = len(tss)
+		}
+		wg.Add(1)
+		go func(begin int, end int) {
+			defer wg.Done()
+			reqCopy := &metricsService.ExportMetricsServiceRequest{
+				ResourceMetrics: req.ResourceMetrics[begin:end],
+			}
+
+			var md metadata.MD
+			var err error
+
+			if _, err = service.Export(c.grpcMetadata(ctx), reqCopy, grpc.Trailer(&md)); err != nil {
+				_ = level.Error(c.logger).Log(
+					"msg", "export failure",
+					"err", truncateErrorString(err),
+					"size", proto.Size(reqCopy),
+					"trailers", fmt.Sprint(md),
+					"recoverable", isRecoverable(err),
+				)
+				errs <- err
+				return
+			}
+
+			_ = level.Debug(c.logger).Log(
+				"msg", "successful write",
+				"records", end-begin,
+				"size", proto.Size(reqCopy),
+				"trailers", fmt.Sprint(md),
+			)
+		}(i, end)
+	}
+	wg.Wait()
+	close(errs)
+	if err, ok := <-errs; ok {
+		return err
+	}
+	return nil
+}
+
+func (c *Client) Close() error {
+	if c.conn == nil {
+		return nil
+	}
+	return c.conn.Close()
+}
+
+func (c *Client) grpcMetadata(ctx context.Context) context.Context {
+	return metadata.NewOutgoingContext(ctx, c.headers)
+}
+
+// truncateErrorString avoids printing error messages that are very
+// large.
+func truncateErrorString(err error) string {
+	tmp := fmt.Sprint(err)
+	if len(tmp) > maxErrorDetailStringLen {
+		tmp = fmt.Sprint(tmp[:maxErrorDetailStringLen], " ...")
+	}
+	return tmp
+}
+
```

Comment:
Please do not truncate error-messages on this level. If (at all) this should be done at the plugin level.

Reply:
Removed.